Add protoc insertion points for the imports and the end of the file.

awpr opened the pull request:
This allows other plugins to add their own imports and declarations to the modules generated by proto-lens-protoc.

Fixes #32.

judah commented on proto-lens-protoc/src/Generate.hs:
```diff
@@ -105,6 +109,25 @@ generateModule modName imports syntaxType definitions importedEnv
+-- - a no-op TH splice or quasiquoter expression
+insertionPoint :: String -> Decl
+insertionPoint nm = PatBind noLoc PWildCard
+    (UnGuardedRhs (Lit (String ("@@protoc_insertion_point("++nm++")"))))
```
Spaces around the ++'s to separate them from the string constant?

awpr commented:
Fixed the (++) formatting.  Also added an insertion point in the module header for adding LANGUAGE / OPTIONS pragmas.

I think it may be impossible to add a (non-comment) insertion point for the export list (although luckily there currently is no export list).  So this approach will fall over if we ever start generating export lists.

An alternative approach is to re-parse the pretty-printed module into an AST annotated with SrcSpanInfo, then traverse it to generate a list of Comments forming the insertion points and re-print it with exactPrint.  This would be slower but less hacky, and wouldn't rely on insane things like OPTIONS pragmas for nonexistent tools and nameless string literal decls.  Thoughts?

judah commented:
If we re-parsing the file and adding comments in the middle, does that invalidate parsed source locations for declarations that happen later on in the program than the inserted comments?

I'm wondering if we're just running into fundamental limits of haskell-src-exts, and it would be simpler to give up on haskell-src-exts pretty-printing the whole file.  That is:

- change `generateModule` to return a `Text.PrettyPrint.Doc` instead of a `Module`
- use [prettyPrim](http://hackage.haskell.org/package/haskell-src-exts-1.17.1/docs/Language-Haskell-Exts-Pretty.html#v:prettyPrim) to pretty-print the individual parts of a file like `Decl` or `ExportSpec`
- Use `Text.PrettyPrint` combinators to glue the parts together, and intersperse comments with the insertion points.

Up to you; OK to punt on it if you think that would be too much work for now.

awpr commented:
Discussed offline -- this is getting messy and the intended use-case is better served by a separate module anyway, so I'm dropping this for now.
